diesel/src/sqlite/types/json.rs

```diff
@@ -0,0 +1,852 @@
+        0x0C => {
+            if bytes.len() < 2 {
+                return Err("Invalid JSONB data: insufficient bytes for payload size".into());
+            }
+            (bytes[1] as usize, 2) // 1 additional byte for payload size
+        }
+        0x0D => {
+            if bytes.len() < 3 {
+                return Err("Invalid JSONB data: insufficient bytes for payload size".into());
+            }
+            (u16::from_be_bytes([bytes[1], bytes[2]]) as usize, 3) // 2 additional bytes
+        }
+        0x0E => {
+            if bytes.len() < 5 {
+                return Err("Invalid JSONB data: insufficient bytes for payload size".into());
+            }
+            (
+                u32::from_be_bytes([bytes[1], bytes[2], bytes[3], bytes[4]]) as usize,
+                5,
+            ) // 4 additional bytes
+        }
+        0x0F => {
+            if bytes.len() < 9 {
+                return Err("Invalid JSONB data: insufficient bytes for payload size".into());
+            }
+            (
+                usize::try_from(u64::from_be_bytes([
+                    bytes[1], bytes[2], bytes[3], bytes[4], bytes[5], bytes[6], bytes[7], bytes[8],
+                ]))
+                .map_err(Box::new)?,
+                9,
+            ) // 8 additional bytes
+        }
+        _ => return Err("Invalid payload size hint".into()),
+    };
+
+    let total_size = header_size + payload_size;
+    if bytes.len() < total_size {
+        return Err(format!(
+            "Invalid JSONB data: insufficient bytes for value of type {}, expected {} bytes, got {}",
+            element_type,
+            total_size,
+            bytes.len()
+        )
+        .into());
+    }
+
+    let payload_bytes = &bytes[header_size..total_size];
+
+    let value = match element_type {
+        JSONB_NULL => Ok(serde_json::Value::Null),
+        JSONB_TRUE => Ok(serde_json::Value::Bool(true)),
+        JSONB_FALSE => Ok(serde_json::Value::Bool(false)),
+        JSONB_INT => read_jsonb_int(payload_bytes, payload_size),
+        JSONB_INT5 => Err("INT5 is not supported".into()),
+        JSONB_FLOAT => read_jsonb_float(payload_bytes, payload_size),
+        JSONB_FLOAT5 => Err("FLOAT5 is not supported".into()),
+        JSONB_TEXT => read_jsonb_text(payload_bytes, payload_size),
+        JSONB_TEXTJ => read_jsonb_textj(payload_bytes, payload_size),
+        JSONB_TEXTRAW => Err("TEXTRAW is not supported".into()),
+        JSONB_TEXT5 => Err("TEXT5 is not supported".into()),
+        JSONB_ARRAY => read_jsonb_array(payload_bytes, payload_size),
+        JSONB_OBJECT => read_jsonb_object(payload_bytes, payload_size),
+        _ => Err(format!("Unsupported or reserved JSONB type: {}", element_type).into()),
+    }?;
+
+    Ok((value, total_size))
+}
+
+// Read a JSON integer in canonical format (INT)
+fn read_jsonb_int(bytes: &[u8], payload_size: usize) -> deserialize::Result<serde_json::Value> {
```

Comment:
bytes should already be payload_size here, right? In that case we don't need to take another subslice below. (The same likely applies to the other read functions as well)